### DISCUSSION TITLE
SSL_write() Intermittently Fails to Send -- Apparently due to TCP Retransmission Failures, and SSL_read() on the Receiving end Hangs Indefinitely.

### DISCUSSION THREAD

**Discussion by horanoff** [openssl/openssl#23309] [Q&A]
Hello Kind Developers,

We are compiling OpenSSL 3.2.0 on 3 different platforms { 64-bit Windows, 64-bit Linux (Ubuntu), and ARM }.  We have a server running on Windows 64-bit (it could run on any of the 3 platforms, but in this test it is Windows 64-bit), and multiple clients running on all 3 types of platforms.  I'm experiencing a problem with SSL_write() on the ARM client only, but it is the only client actually on a network (the other clients are tested on the same local machine.  SSL_write() is working fine on both Windows and Linux AMD locally.  When a server issues a JSON-RPC request for a screenshot from a client, the client respond with a JSON-RPC response with a Base64-encoded JPG.  I'm using this screenshot example because of the large amount of data in the response, and is therefore a relatively intense test.  This response has an outer wrapper containing a 32-bit length followed by the JSON request/response data of that length (much like OpenSSL has).  Everything typically works fine on all 3 clients, but on the remote ARM client there's an intermittent problem I'm experiencing in which SSL_read() on the server side hangs and never wakes up mid-transmission (part way through the JSON-RPC response packet loop) and is stuck in SSL_read() trying to read its own payload data after reading its own payload length.  See tls_common.c around line number 676 in version 3.2.0.  It is stuck in the `rret = rl->funcs->read_n(rl, more, more, 1, 0, &n);` function.

```
if (more > 0) {
    /* now rl->packet_length == SSL3_RT_HEADER_LENGTH */

    rret = rl->funcs->read_n(rl, more, more, 1, 0, &n);
    if (rret < OSSL_RECORD_RETURN_SUCCESS)
        return rret;     /* error or non-blocking io */
}
```
I have the ability to disable SSL, and when I do our code speaks directly through Berkeley Sockets.  This works fine on all 3 platforms, and when OpenSSL is used to encrypt that data it works reliably on the other two non-local platforms. 

I believe this is a problem related to SSL_write(), not SSL_read().  By adding additional logging, I see that SSL_write() successfully writes all data blocks of the response -- but presumably, this data simply gets cached and SSL_write() returns promptly.

Using Wireshark to introspect into TCP traffic, I'm seeing a group of _TCP Retransmission_ events around the time of the unwanted event (see the middle/bottom of this Wireshark screenshot).

![image-2024-01-12-11-53-39-615](https://github.com/openssl/openssl/assets/22893499/45413d29-d256-4b96-b58d-d8d2adfa8e2e)

I'm using version 3.2.0, but also found this problem in version 3.1.2 before I updated to see if the problem had been resolved.
I would think this must be happening close to the BIO level, and I don't think there's anything I could do cause _TCP Retransmission_ events from a level above the OpenSSL library.

Any help and/or thoughts would be greatly appreciated!

**Comment by t8m** [comment]
This really does not sound to me like a bug in OpenSSL but rather some issue with the kernel, or machine connectivity or something else in the network stack.

Maybe @mattcaswell or @hlandau could have some idea?

**Reply by horanoff** [reply]
I totally agree with you that it sounds like a connectivity or network stack issue; however, taking OpenSSL out of the equation and talking directly through TCP sockets works just fine.  That's the only thing that is stopping me in my tracks from declaring this a hardware/OS-related issue.  

The are 2 very significant things that dissuade me from concluding this is a bug in OpenSSL:  

1. It sounds like a connectivity or network stack issue.
2. I would imagine that OpenSSL has been **vigorously** tested.
 
Having said that, I was hoping AArch64 might have received less testing, and perhaps less testing with large data blocks.  But even so, I'm not sure how OpenSSL would even have such low-level access to cause these TCP retransmissions.  It definitely sounds like something at a lower level than even the BIO level.

I don't file bug reports like this lightly.  The ONLY thing that brings me here is that everything works fine if OpenSSL is bypassed and I use Berkeley sockets instead.

FYI, I'm using blocking I/O for simplicity, so there shouldn't be any NBIO user errors on my end -- just SSL_read() and SSL_write().

I'll talk to our hardware guy and OS guy to see if they have any additional thoughts.

Thank you for your response!  It's valuable to know you are thinking the same things I have.

**Comment by nhorman** [comment]
based on the output above, it seems like you're encountering a network error of some sort (the root cause is impossible to say with the information given).  I expect some frames are getting dropped somewhere (hence the tcp retransmissions), and as a result BIO_read on the socket is hanging waiting for the kernel to complete reassembly of the TCP stream which is missing a chunk of sequential data.  The goal should be to figure out where the missing data is getting dropped, which will either be on an intermediate network device or somewhere in your kernel.

**Reply by horanoff** [reply]
Thank you for your response.  I'm going to have our hardware guy and OS guy take a look at the issue.  Having said that, I'm still baffled as to why everything works fine when talking directly through Berkeley sockets.

**Comment by tom-cosgrove-arm** [comment]
By "ARM" what exactly do you mean? A 32-bit Armv7 system? A 64-bit AArch64 system? A Raspberry Pi (which can be either 32- or 64-bit) running either 32- or 64-bit Linux? Something other than Linux?

Are you able to supply a simple reproducer client and server (e.g. client makes request from server, server sends a block of malloc()d data of similar size to a screenshot)?

How big are the screenshots? Assuming 32-bit ARM, could that be an issue (though would have to be a very big screenshot!)

**Reply by horanoff** [reply]
Hi Tom, it's a 64-bit AArch64 system running Ubuntu 18.04, and the data being sent is about 1.5 megabytes after base64-encoding.

I will try and capture the pre-encrypted data being sent, and then try to create a reproducible example.  Having said that, this is an intermittent problem and may have more to do with timing than data, or other possibilities of course.  If none of my experience sounds familiar, then I agree this is a good next step.

**Comment by horanoff** [comment]
One oversight on my part is that is that I'll never see TCP retransmissions on my Windows and Linux AMD64 clients because they are VMs local to my development machine.  I'm going to procure actual hardware that I will add to the network for an apples-to-apples comparison.  Although it's true I've only reproduced the problem on ARM64, my title that says "On ARM only" could be quite a mischaracterization.

I have a bit more work to do to narrow this down.  Thank you all for your insights!

**Comment by horanoff** [comment]
It took me weeks, but I was finally able to solve the problem.  It was a multi-thread locking issue.  I did not have confidence that OpenSSL would automatically handle multithreading properly out-of-the-box, and because of that I added mutex locking around both reading and writing in order to serialize these operations; however, this intermittent problem was only resolved by replacing my own mutex locking mechanism with OpenSSL's locking mechanism -- an example of which I (eventually) found on the internet:
```
    static CRYPTO_ONCE once = CRYPTO_ONCE_STATIC_INIT;
    static CRYPTO_RWLOCK *lock;

    static void myinit(void)
    {
        lock = CRYPTO_THREAD_lock_new();
    }

    static int mylock(void)
    {
        if (!CRYPTO_THREAD_run_once(&once, myinit) || lock == NULL)
            return 0;
        return CRYPTO_THREAD_write_lock(lock);
    }

    static int myunlock(void)
    {
        return CRYPTO_THREAD_unlock(lock);
    }
```
Clearly, OpenSSL[/Crypto] is doing something under the hood that requires the user to use this specific locking mechanism.  It's amazing to me that OpenSSL leaves it to the developer to handle this locking at all.  At the very least, it is fair to say that this API is _unusual_.  It would be trivial for OpenSSL to provide an API that properly handles multithreading for the user without exposing internal mechanisms like this.  In fact, I ended up wrapping this functionality in a higher-level API myself to solve this problem seamlessly.  That was not rocket science, and in my humble opinion, a missed opportunity for OpenSSL.  I highly recommend OpenSSL take the time to address this missed opportunity.  I see a lot of people on the internet struggling with the unclear documentation and lack of examples as it relates to getting OpenSSL to work reliably in a multi-threaded environment.  There's a lot of good will to be gained here.

Thank you @t8m, @nhorman, and @tom-cosgrove-arm for your assistance in helping me hunt this down!  Although this was one of the most frustrating experiences in my 40-year career, in the end I got this working thanks partly to your assistance.

**Reply by t8m** [reply]
So your application called SSL_* API calls on a single SSL object simultaneously from multiple threads? That is not supported, you have to ensure such calls from multiple threads are serialized by locking/mutexes.

There is the openssl-threads manual page: https://www.openssl.org/docs/man3.0/man7/openssl-threads.html Unfortunately that page does not mention SSL objects. As adding locks/mutexes around every SSL_ API call would increase the overhead for applications that do their own locking or do not use SSL objects simultaneously from multiple threads we cannot add the locks there.

**Reply by t8m** [reply]
I've created https://github.com/openssl/openssl/issues/23446

**Reply by horanoff** [reply]
> So your application called SSL_* API calls on a single SSL object simultaneously from multiple threads? That is not supported, you have to ensure such calls from multiple threads are serialized by locking/mutexes.

Thanks @t8m.  No, I was not performing those operations simultaneously.  I was using a mutex -- but as others discovered, it's not sufficient to use your own mutex.  It appears you must use SSL's mutexes.  That did appear to solve the problem in my case.

**Reply by mattcaswell** [reply]
> it's not sufficient to use your own mutex. It appears you must use SSL's mutexes. 

That's just strange. I can't think of anything in OpenSSL that would cause that to be the case.

**Reply by t8m** [reply]
That seems very strange. Any mutex should work, if it works properly - i.e., serializes the access. That sounds like a problem with the mutex implementation. What exact mutex was used that did not work? I do not know if there are any pecularities in regards to mutexes on Windows though.

**Reply by horanoff** [reply]
It seems pretty clear that some combination of the mutex switch and the printfs I inserted changed the timing causing things to work, but when I remove the printfs, then the intermittent problem resurfaces.  I then realized that if I simply insert a sleep statement between write operations, the problem goes away.  I tried this on different hardware and I still see the same problem.  I did not try a different network switch, but it seems pretty clear to me that this is a software issue.  I've replied previously below with my sample code.  This sample code executes on both client and server, and there are no other SSL or socket operations happening outside of this code.  With that in mind, I don't see how the bug can exist outside of my sample code, or outside of OpenSSL itself.  The only other thing I can think of that might cause this issue would be a SSL/socket configuration issue.  Calling "SSL_write() too quickly" shouldn't be a condition that should cause this issue, but that seems to be the case.  If there's a check I need to make before calling SSL_write(), then I must have missed this in the documentation.

> As adding locks/mutexes around every SSL_ API call would increase the overhead for applications that do their own locking or do not use SSL objects simultaneously from multiple threads we cannot add the locks there.

Berkeley sockets are generally reentrant in that you can send data and receive data asynchronously from different threads without the need for a locking mechanism.  The only thing you need to serialize **_separately_** are simultaneous read operations and simultaneous write operations.  Does OpenSSL support separate SSL contexts for reading and writing?  In other words, is it possible to assign multiple SSL pointers to a single socket -- one for reading and one for writing using ::SSL_set_fd(ssl, berkeleySocket)?  If that's not possible, then OpenSSL can't be as efficient as other solutions that properly handle asynchronous bidirectional connections reentrantly.

**Comment by horanoff** [comment]
I believe you are both right.  I believe that by changing my Mutex implementation and/or my inserting printfs, it changed the timing such that it made things work much better, but in the end it didn't solve my problem. 

I have concluded (hopefully correctly) that the best implementation for bidirectional communications with OpenSSL is to use non-blocking I/O and to have a single thread handle both SSL_read() and SSL_write() for every socket connection (every SSL instance) -- or at least this is the approach I'm taking to narrow down the problem.  Hopefully this allows me to avoid using mutexes altogether.  I have altered my code to do just that, and although this simplified the code, it did not eliminate the problem. 

When a large amount of data needs to be sent through the socket (like my screenshots I'm sending back that are Base64-encoded), multiple SSL_write() operations must be executed in a loop since each SSL_write() operation returns that it is only writing 16384 bytes per call (as expected).  I continue to call SSL_write() until all of the data has been written.  If I perform those SSL_write() operations in a tight loop, and then loop back around to SSL_read(), then SSL_read() will often return SSL_ERROR_SSL "error:00000001:lib(0)::reason(1)".  However, if I insert a 10 millisecond delay after each SSL_write() operation, then when I loop back around to SSL_read() then everything seems to work reliably (although more slowly). 

Below is my thread that serializes both read and write operations.  Notice near the bottom the 10 millisecond sleep between SSL_write() operations.  If I remove this sleep statement, then things go awry.  If I reduce the 10 millisecond delay down to about 4 milliseconds, then things begin to become unreliable.  There aren't any examples I could find for how to properly perform bidirectional communications using OpenSSL, so I'm flying a little blind here.  Please let me know if there's something I'm missing in my code in order to make this work properly.  Thank you!

```
int64_t SCA::SecureSocketMonitorTask::_main()
{
    // Retrieve the SSL pointer from the security context.
    auto ssl = (SSL *)this->mSocketSecurity->getContext()->shared_ptr<SCA::SecureSocketContext>()->ssl;

    // Grab the Berkeley socket from the parent socket instance.
    auto berkeleySocket = dynamic_cast<SCA::Socket *>(this->mSocketSecurity)->mBerkeleySocket;

    // Keep running until the connection is disconnected.
    while (!this->abortSignal.readSignal())
    {
        // See if there's anything to read.  Per the specification, this is the maximum read block size for SSL.
        char buffer[16384];
        auto result = SSL_read(ssl, buffer, sizeof(buffer));

        // If we were not successful, then find out why.
        if (result == -1)
        {
            fd_set fds;
            FD_ZERO(&fds);
            FD_SET(berkeleySocket, &fds);
            struct timeval timeout = { 0, 100 * 1000 };
            switch (auto sslError = ::SSL_get_error(ssl, result))
            {
                case SSL_ERROR_WANT_READ:  ::select(FD_SETSIZE, &fds, NULL, NULL, &timeout); break;
                case SSL_ERROR_WANT_WRITE: ::select(FD_SETSIZE, NULL, &fds, NULL, &timeout); break;
                case SSL_ERROR_SSL:
                case SSL_ERROR_SYSCALL:
                default:
                {
                    Scope("While calling SSL_read().");
                    SslError(ssl);

                    // Create a new message.
                    auto message = std::make_shared<SecureMessage>();

                    // Set the signal that says our error state has been cleared, end then
                    // set the error state.
                    message->error = thisTask.exportError();

                    // Push the message onto the queue.
                    this->readQueue.push(message);

                    // Set the signal that indicates there's something to do.
                    if (auto tcpSocket = SCA::shared_ptr<SCA::TcpSocket>(this->mSecureSocket))
                        tcpSocket->setSignal();

                    // All done handling this read error.
                    break;
                }
            }
        }

        // Otherwise, we were successful if our result is a positive integer.
        else if (result > 0)
        {
            // Create a new message.
            auto message = std::make_shared<SecureMessage>();

            // Write the decrypted data into the message.
            message->writeMessage(buffer, result);

            // Reset the offset to 0 in preparation for the consumer to read
            // the message.
            message->setOffset(0);

            // Push the message onto the queue.
            this->readQueue.push(message);

            // Set the signal that indicates there's something to do.
            if (auto tcpSocket = SCA::shared_ptr<SCA::TcpSocket>(this->mSecureSocket))
                tcpSocket->setSignal();
        }

        // Is there anything to write through the encrypted socket?
        if (auto request = this->writeQueue.pull())
        {
            // Since we know OpenSSL will only write a maximum of 16384 bytes,
            // there's no sense trying to 
            while (  request->getOffset() 
                   < request->getLength())
            {
                // Attempt to SSL_write() data.
                auto result = SSL_write(ssl, 
                                        request->getPointer() + request->getOffset(), 
                                        request->getLength()  - request->getOffset());

                // If we were not successful, then find out why.
                if (result == -1)
                {
                    fd_set fds;
                    FD_ZERO(&fds);
                    FD_SET(berkeleySocket, &fds);
                    struct timeval timeout = { 0, 100 * 1000 };
                    auto sslError = ::SSL_get_error(ssl, result);
                    switch (sslError)
                    {
                        case SSL_ERROR_WANT_READ:  ::select(FD_SETSIZE, &fds, NULL, NULL, &timeout); break;
                        case SSL_ERROR_WANT_WRITE: ::select(FD_SETSIZE, NULL, &fds, NULL, &timeout); break;
                        case SSL_ERROR_SSL:
                        case SSL_ERROR_SYSCALL:
                        default:
                        {
                            Scope("While calling SSL_write().");
                            SslError(ssl);

                            // Export the error as the Secure Message's response.
                            request->error = thisTask.exportError();

                            // All done handling this request.
                            request->setOffset(request->getLength());
                            break;
                        }
                    }
                }

                // Otherwise, we were successful.  Adjust our pointers to point to the
                // next data block to write.
                else
                {
                    // Set the new data offset to just past the data we successfully
                    // wrote using SSL_write().
                    request->setOffset(request->getOffset() + result);

                    // This delay appears to be necessary in order to avoid
                    // an intermittent SSL_read() error (?)
//                    SCA::Platform::sleep(10);
                }
            }

            // The request has completed.
            request->setSignal();
        }
    }

    // This return value isn't used.
    return 0;
}
```
